### PR TITLE
[PDN] Power switch API update

### PR DESCRIPTION
**Pull request by minghungumich**
Update set_voltage_domain, define_pdn_grid, add_global_connection and create insert_power_switches to place switched power domain and selected power switch. Testcase are provided in sky130hd. It replaces VDD rail with VDD_SW, insert VDD_SW strap near VDD, and place power switch instances on the intersection. It automatically add via for power switch connection and move the tap cell to prevent overlapping.
截圖 2022-04-01 18 39 49

Signed-off-by: minghungumich <minghung@umich.edu>

### CODE REVIEW

**Comment by minghungumich** [comment]
I think my old fork is too out of date and will have merge issue. So I decided to close the original PR [#1579](https://github.com/The-OpenROAD-Project/OpenROAD/pull/1579).
Now the tap cell movement is implemented well and all of change is put on a new PR [#1744](https://github.com/The-OpenROAD-Project/OpenROAD/pull/1744).
Please help me add the reviewer. Thanks!

**Comment by Colin-Holehouse** [comment]
The code has been updated to the latest OR code, but TCL code to insert power switches is no longer active.

**Comment by arlpetergadfort** [comment]
> The code has been updated to the latest OR code, but TCL code to insert power switches is no longer active.

@Colin-Holehouse once it is in, I'll migrate it over to the C++ version so the functionality in TCL should only be missing for a little while.

File: src/pdn/src/pdn.tcl [hunk 1]
```diff
@@ -254,6 +266,54 @@ proc define_pdn_grid {args} {
...
+
+proc define_power_switch_cell {args} {
+  pdngen::check_design_state
+
+  sta::parse_key_args "define_power_switch_cell" args \
+    keys {-name -control -acknowledge -power_switchable -power -ground} 
+
+  if {[llength $args] > 0} {
+    utl::error PDN 1182 "Unexpected argument [lindex $args 0] for define_power_switch_cell command."
+  }
```
Comment (arlpetergadfort) [inline src/pdn/src/pdn.tcl:284]: Please mirror the other commands like `add_global_connection`:

```
sta::parse_key_args "add_global_connection" args \
    keys {-net -inst_pattern -pin_pattern} \
    flags {-power -ground -defer_connection}

  sta::check_argc_eq0 "add_global_connection" $args

  pdn::check_design_state "add_global_connection"
```
This way it's consistent with all the other calls.

File: src/pdn/src/pdn.tcl [hunk 2]
```diff
@@ -254,6 +266,54 @@ proc define_pdn_grid {args} {
...
+    utl::error PDN 1183 "The -name argument is required."
+  }
+
+  if {![info exists keys(-control)]} {
+    utl::error PDN 1184 "The -control argument is required."
+  }
+
+  if {![info exists keys(-acknowledge)]} {
+    utl::error PDN 1185 "The -acknowledge argument is required."
+  }
```
Comment (arlpetergadfort) [inline src/pdn/src/pdn.tcl:296]: Is this required? is it possible that there is no acknowledge pin?

File: src/pdn/src/pdn.tcl [hunk 3]
```diff
@@ -254,6 +266,54 @@ proc define_pdn_grid {args} {
...
+
+  if {![info exists keys(-power_switchable)]} {
+    utl::error PDN 1186 "The -power_switchable argument is required."
+  }
```
Comment (arlpetergadfort) [inline src/pdn/src/pdn.tcl:300]: What does this define? If it's the net, it should get this from the domain its attached so as to avoid duplicating information.

File: src/pdn/src/pdn.tcl [hunk 4]
```diff
@@ -254,6 +266,54 @@ proc define_pdn_grid {args} {
...
+    utl::error PDN 1187 "The -power argument is required."
+  }
+
+  if {![info exists keys(-ground)]} {
+    utl::error PDN 1188 "The -ground argument is required."
+  }
+
+  if {[llength $args] > 0} {
+    utl::error PDN 1189 "Unrecognized argument [lindex $args 0] for define_power_switch_cell."
+  }
```
Comment (arlpetergadfort) [inline src/pdn/src/pdn.tcl:312]: This is not possible, so please remove

File: src/pdn/src/pdn.tcl [hunk 5]
```diff
@@ -254,6 +266,54 @@ proc define_pdn_grid {args} {
...
+
+  pdngen::define_power_switch_cell {*}[array get keys]
```
Comment (arlpetergadfort) [inline src/pdn/src/pdn.tcl:314]: This will be changed to call into the CPP version

File: src/pdn/test/power_switch.api.tcl [hunk 6]
```diff
@@ -0,0 +1,52 @@
+source "helpers.tcl"
+
+read_lef ../../../test/sky130hd/sky130hd.tlef 
+read_lef ../../../test/sky130hd/sky130_fd_sc_hd_merged.lef
```
Comment (arlpetergadfort) [inline src/pdn/test/power_switch.api.tcl:4]: Why did you delete the sky130hd link? Please restore and fix all the test to use that instead of `../../../test/`

File: src/pdn/test/power_switch.api.tcl [hunk 7]
```diff
@@ -0,0 +1,52 @@
+source "helpers.tcl"
```
Comment (arlpetergadfort) [inline src/pdn/test/power_switch.api.tcl:1]: Add a comment to clarify exactly what this test is checking.

File: src/pdn/test/power_switch.daisy.api.tcl [hunk 8]
```diff
@@ -0,0 +1,52 @@
+source "helpers.tcl"
```
Comment (arlpetergadfort) [inline src/pdn/test/power_switch.daisy.api.tcl:1]: Add a comment to clarify exactly what this test is checking.

File: src/pdn/test/power_switch.api.tcl [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+read_lef ../../../test/sky130hd/sky130hd.tlef 
+read_lef ../../../test/sky130hd/sky130_fd_sc_hd_merged.lef 
+
+read_lef sky130hd/power_switch.lef 
+
+read_def power_switch/2_5_floorplan_tapcell.def
```
Comment (arlpetergadfort) [inline src/pdn/test/power_switch.api.tcl:8]: to be consistent, please rename to sky130_power_switch/floorplan.def (name should make it clear the process)

File: src/pdn/test/power_switch.api.tcl [hunk 10]
```diff
@@ -0,0 +1,52 @@
+source "helpers.tcl"
+
+read_lef ../../../test/sky130hd/sky130hd.tlef 
+read_lef ../../../test/sky130hd/sky130_fd_sc_hd_merged.lef 
+
+read_lef sky130hd/power_switch.lef
```
Comment (arlpetergadfort) [inline src/pdn/test/power_switch.api.tcl:6]: please move to sky130_power_switch/ (keep test files together)

File: src/pdn/test/region_non_rect.ok [hunk 11]
```diff
@@ -1,10 +1,10 @@
-[INFO ODB-0222] Reading LEF file: sky130hd/sky130hd.tlef
```
Comment (arlpetergadfort) [inline src/pdn/test/region_non_rect.ok:1]: test should not have changed.

File: src/pdn/test/region_non_rect.tcl [hunk 12]
```diff
@@ -1,8 +1,8 @@
 # test for error at non rectangular region
 source "helpers.tcl"
 
-read_lef sky130hd/sky130hd.tlef 
-read_lef sky130hd/sky130_fd_sc_hd_merged.lef 
+read_lef ../../../test/sky130hd/sky130hd.tlef
```
Comment (arlpetergadfort) [inline src/pdn/test/region_non_rect.tcl:4]: fix

File: src/pdn/test/region_temp_sensor.ok [hunk 13]
```diff
@@ -1,10 +1,10 @@
-[INFO ODB-0222] Reading LEF file: sky130hd/sky130hd.tlef
+[INFO ODB-0222] Reading LEF file: ../../../test/sky130hd/sky130hd.tlef
```
Comment (arlpetergadfort) [inline src/pdn/test/region_temp_sensor.ok:1]: fix

File: src/pdn/test/region_temp_sensor.tcl [hunk 14]
```diff
@@ -1,8 +1,8 @@
 # test for set_voltage_domain -region
 source "helpers.tcl"
 
-read_lef sky130hd/sky130hd.tlef 
-read_lef sky130hd/sky130_fd_sc_hd_merged.lef 
+read_lef ../../../test/sky130hd/sky130hd.tlef
```
Comment (arlpetergadfort) [inline src/pdn/test/region_temp_sensor.tcl:4]: fix

File: src/pdn/test/regression_tests.tcl [hunk 15]
```diff
@@ -77,3 +77,18 @@ record_tests {
 
   existing
 }
+
+set extra_tests {
```
Comment (arlpetergadfort) [inline src/pdn/test/regression_tests.tcl:81]: What does this do? most of those files are no longer present, so the only ones that should be there is power_switch* and those should be active in the actual regression tests (even if they are using the legacy calls)

File: src/pdn/test/sky130hd/power_switch.lef [hunk 16]
```diff
@@ -0,0 +1,55 @@
+VERSION 5.6 ;
```
Comment (arlpetergadfort) [inline src/pdn/test/sky130hd/power_switch.lef:1]: move to sky130_power_switch

**Review by arlpetergadfort — CHANGES_REQUESTED** [review]

File: src/pdn/doc/define_power_switch_cell.md [hunk 17]
```diff
@@ -0,0 +1,34 @@
+# define_power_switch_cell
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/doc/define_power_switch_cell.md:1]: File removed

File: src/pdn/README.md [hunk 18]
```diff
@@ -65,12 +93,14 @@ The -region argument specifies the name of a region of the design. This region m
...
 
+If the voltage domain is a switched power domain, then the name of the swiched power net must be specified with the -switched_power option.
 ```
 set_voltage_domain [-name name] \
                    -power power_net \
                    -ground ground_net \
                    [-region region_name] \
-                   [-secondary_power secondary_power_net]
+                   [-secondary_power secondary_power_net] \
+                   [-switched_power <switched_power_net>] \
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/README.md:103]: Done

File: src/pdn/include/pdn/PdnGen.hh [hunk 19]
```diff
@@ -121,6 +121,8 @@ class PdnGen
                                odb::dbNet* ground,
                                const std::vector<odb::dbNet*>& secondary_nets,
                                odb::dbRegion* region);
+  void setVoltageDomainSwitchedPower(const std::string& name,
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/include/pdn/PdnGen.hh:124]: Done

File: src/pdn/src/PdnGen.cc [hunk 20]
```diff
@@ -486,6 +486,22 @@ void PdnGen::makeRegionVoltageDomain(
...
+  for (const auto& domain: domains_) {
+    if (domain->getName() == name) {
+      domain->setSwitchedPower(switched_power);
+      return;
+    }
+  }
+  logger_->error(utl::PDN,
+		 195,
+		 "Voltage domain {} does not exist.",
+		 name);
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/src/PdnGen.cc:502]: Removed error.

File: src/pdn/src/PdnGen.i [hunk 21]
```diff
@@ -158,6 +158,12 @@ void make_region_domain(const char* name, odb::dbNet* power, odb::dbNet* ground,
   pdngen->makeRegionVoltageDomain(name, power, ground, secondary_nets, region);
 }
 
+void set_domain_switched_power(const char* name, odb::dbNet* switched_power)
+{
+  PdnGen* pdngen = ord::getPdnGen();
+  pdngen->setVoltageDomainSwitchedPower(name, switched_power);
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/src/PdnGen.i:164]: Done.

File: src/pdn/src/pdn.tcl [hunk 22]
```diff
@@ -219,6 +220,15 @@ proc set_voltage_domain {args} {
   } else {
     pdn::make_region_domain $name $pwr $gnd $secondary $region
   }
+
+  if {[info exists keys(-switched_power)]} {
+    set switched_power_net_name $keys(-switched_power)
+    set db_net [[ord::get_db_block] findNet $switched_power_net_name]
+    if {$db_net == "NULL"} {
+      set db_net [odb::dbNet_create [ord::get_db_block] $switched_power_net_name]
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/src/pdn.tcl:228]: Yes - made it special.

File: src/pdn/src/pdn.tcl [hunk 23]
```diff
@@ -254,6 +266,54 @@ proc define_pdn_grid {args} {
   }
 }
 
+sta::define_cmd_args "define_power_switch_cell" {[-name <name>] \
+                                                 [-control <control_name>] \
+                                                 [-acknowledge <acknowledge_name>] \
```
[the remark this answers is not in the document]
Comment (Colin-Holehouse) [reply src/pdn/src/pdn.tcl:271]: Yes, these are pins - update to control_pin and acknowledge_pin. The same is true for power_switchable/power and ground - these are all also pins of the cell.

File: src/pdn/src/pdn.tcl (hunk 1 above)
Comment (Colin-Holehouse) [reply src/pdn/src/pdn.tcl:284]: Done.

File: src/pdn/src/pdn.tcl (hunk 2 above)
Comment (Colin-Holehouse) [reply src/pdn/src/pdn.tcl:296]: Agreed - acknowledge is optional

File: src/pdn/src/pdn.tcl (hunk 3 above)
Comment (Colin-Holehouse) [reply src/pdn/src/pdn.tcl:300]: This defines the pin on the power switch net that outputs the switched version of the input power pin.
Using the USE POWER attribute on the cell isnt sufficient to distinguish between the two power pins on the cell.

File: src/pdn/test/power_switch.api.tcl (hunk 6 above)
Comment (Colin-Holehouse) [reply src/pdn/test/power_switch.api.tcl:4]: Ah - didnt catch on that this was a link. Fixed
Since the power_swtich.lef is a made up cell for the purposes of creating a testcase, I've moved this to the power_switch directory.

File: src/pdn/test/power_switch.api.tcl (hunk 7 above)
Comment (Colin-Holehouse) [reply src/pdn/test/power_switch.api.tcl:1]: Done

Comment (Colin-Holehouse) [reply src/pdn/test/power_switch.api.tcl:1]: Done

File: src/pdn/test/power_switch.daisy.api.tcl (hunk 8 above)
Comment (Colin-Holehouse) [reply src/pdn/test/power_switch.daisy.api.tcl:1]: Done

File: src/pdn/test/power_switch.api.tcl (hunk 9 above)
Comment (Colin-Holehouse) [reply src/pdn/test/power_switch.api.tcl:8]: Done

File: src/pdn/test/power_switch.api.tcl (hunk 10 above)
Comment (Colin-Holehouse) [reply src/pdn/test/power_switch.api.tcl:6]: Done

File: src/pdn/test/region_non_rect.tcl (hunk 12 above)
Comment (Colin-Holehouse) [reply src/pdn/test/region_non_rect.tcl:4]: Fixed

File: src/pdn/test/sky130hd/power_switch.lef (hunk 16 above)
Comment (Colin-Holehouse) [reply src/pdn/test/sky130hd/power_switch.lef:1]: Done

File: src/pdn/test/region_temp_sensor.ok (hunk 13 above)
Comment (Colin-Holehouse) [reply src/pdn/test/region_temp_sensor.ok:1]: Fixed

File: src/pdn/test/regression_tests.tcl (hunk 15 above)
Comment (Colin-Holehouse) [reply src/pdn/test/regression_tests.tcl:81]: Yes, these tests were removed, are you saying that the replacement tests fully cover everything that was being checked before? Or are you saying that these aren't really unit tests, but are more like integration tests?

Clearly the power_switches tests need to exist, but with the presence of the C++ code, isnt this code now orphaned? Should I expect to be able to run pdngen with only the TCL code implementation and bypass the C++ code (on a temporary basis). I dont see how I can do that when the TCL commands only seed the C++ code, not the TCL code.

If the test isnt going to run cleanly, then the CI checks will fail. I dont want to checkin the results of the failure just to get a clean CI run.

File: src/pdn/test/region_non_rect.ok (hunk 11 above)
Comment (Colin-Holehouse) [reply src/pdn/test/region_non_rect.ok]: Fixed

File: src/pdn/test/regression_tests.tcl (hunk 15 above)
Comment (arlpetergadfort) [reply src/pdn/test/regression_tests.tcl]: @Colin-Holehouse the tests that are there only handle the c++ code, since the plan is to deprecate the TCL code eventually (but keeping it for a while to allow for some transition). I think you could add the power_switch tests that use the TCL code into the regression (those should not fail and will only generate a warning, which is okay to ignore), that way they are part of the tests and once the C++ code is added to handle the power_switches then those tests can be converted to use the c++ code.  The CFG file can still be provided to setup the TCL pdngen. ie. `pdngen power_switch.cfg` will trigger the TCL code version to run.

Comment (Colin-Holehouse) [reply src/pdn/test/regression_tests.tcl]: Ok - thanks, I can get the power_switch tests to run using the base TCL code without going through the C++ code base.

File: src/pdn/src/pdn.tcl (hunk 5 above)
Comment (Colin-Holehouse) [reply src/pdn/src/pdn.tcl:305]: Yes - I assume that you will handle that as part of your process for incorporating the power switch insertion TCL code.

**Review by arlpetergadfort — APPROVED** [review]
Looks good me to.
